### DISCUSSION TITLE
Interface() for async operations

### DISCUSSION THREAD

**Discussion by slaff** [jfjlaros/cpp-simple-rpc#5] [Ideas]
I have noticed that the simpleRPC library expects a function and all of its parameters to be in the incoming stream. For a framework like Sming that uses an asynchronous TCP operations that might not always be the case. 

Let's say we have the following function
```C++
uint32_t sum(uint16_t a, uint8_t b)
```

A client using Tcp should send one byte for the id, let's say 1, and 3 more bytes - two for parameter `a` and one for parameter `b`. Or 4 bytes in total. Using async TCP it might happen that the first 2 bytes are part of one tcp packet and the next 2 are part of the following tcp packet. Would it be possible to improve the `interface()` function to handle that? For example for every exported function a quick calculation of the needed bytes for parameters can be run. If the total length of the parameter data cannot be determined then the expected size should be -1.  The value can be stored and after that used in `interface` to check. 
if a RCP method is called the byte for the method id should be consumed. After that if the needed bytes for this method are zero or positive number then a check should be performed to see if the available bytes in the stream >= needed bytes.
If there are enough available bytes then they should be consumed. And the interface function should return the size of the consumed bytes or just 0. Other the `interface()` should return `-1` specifying that more bytes are needed without consuming the stream bytes.  

What do you think?

**Comment by jfjlaros** [comment]
I think we handle this by using a serial protocol over TCP. The underlying library blocks until all data is received.

**Reply by jfjlaros** [reply]
You might want to have a look at [this](https://github.com/jfjlaros/simpleRPC/blob/master/examples/esp32/esp32.ino) demo. I have used WireShark to confirm that not all data is transmitted in one TCP package.
